**#PyTorch Mini Project (Dataset-Independent)**
Choose One Dataset

Students must choose one dataset from the following list and build a PyTorch ANN model.

1️⃣ MNIST Handwritten Digit Dataset

Digits 0–9 classification.

Dataset link:
https://www.kaggle.com/datasets/hojjatk/mnist-dataset

2️⃣ Fashion-MNIST Dataset

Clothing image classification.

Dataset link:
https://www.kaggle.com/datasets/zalando-research/fashionmnist

3️⃣ Iris Dataset

Flower species classification.

Dataset link:
https://www.kaggle.com/datasets/uciml/iris

4️⃣ Breast Cancer Wisconsin Dataset

Binary classification (malignant vs benign).

Dataset link:
https://www.kaggle.com/datasets/uciml/breast-cancer-wisconsin-data

5️⃣ Titanic Survival Dataset

Predict whether a passenger survived.

Dataset link:
https://www.kaggle.com/datasets/yasserh/titanic-dataset

# Topics Students Must Use

Students must apply the following PyTorch concepts in their project:

Why PyTorch

Tensors

Autograd

PyTorch Training Pipeline

torch.nn Module & torch.optim Module

Dataset and DataLoader

Building ANN using PyTorch

Model training and evaluation



# Project using irish dataset

3️⃣ Iris Dataset

Flower species classification.

Dataset link: https://www.kaggle.com/datasets/uciml/iris


# Question 1

Load the dataset you selected and perform basic preprocessing.

Explain:

number of samples

number of features

target variable

train-test split

In [361]:
import pandas as pd
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from torch.utils.data import DataLoader, Dataset
import matplotlib.pyplot as plt


torch.manual_seed(42)

In [362]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

cpu


In [363]:
df = pd.read_csv("/content/Iris.csv")
df.head()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa


In [364]:
print(df.info())
print(df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB
None
Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64


In [365]:
df.shape

(150, 6)

In [366]:
df.drop(columns=["Id"], inplace=True)
df.head()

,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,5.1,3.5,1.4,0.2,Iris-setosa
1,4.9,3.0,1.4,0.2,Iris-setosa
2,4.7,3.2,1.3,0.2,Iris-setosa
3,4.6,3.1,1.5,0.2,Iris-setosa
4,5.0,3.6,1.4,0.2,Iris-setosa


In [367]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df["Species"] = le.fit_transform(df["Species"])
df.head()

,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


# Question 2

Convert the dataset features into PyTorch tensors.

Explain:

what a tensor is

tensor shape of your dataset

why tensors are required in PyTorch

#Write answer of Question 2

In [368]:
X = df.drop('Species', axis=1).values
y = df['Species'].values

X_tensor = torch.tensor(X, dtype=torch.float32)
y_tensor = torch.tensor(y, dtype=torch.long)

print(X_tensor.shape)
print(y_tensor.shape)

torch.Size([150, 4])
torch.Size([150])


In [369]:
num_classes = len(torch.unique(y_tensor))
print(f"Number of classes in the dataset: {num_classes}")

Number of classes in the dataset: 3


A tensor is a datastructure like NumPy arrays in python which is used to store multi-dimensional matrix with numerical data.

in this dataset the X_tensor shape is 150,4 where 150 is the rows representing the samples and 4 is the columns for number of features. The y tensor shape 150 is the 1D array as the class labels.

Tensors are required in python because :



1.   Neural networks input and output tensors
2.   Tensor provides fast and efficient operations using GPU.

3. for automatic differentiation using autograd we need tensor
4. tensor provides a rich set of mathmetical functions for complex calculation.



# Question 3

Create a simple example using Autograd:

define a tensor with requires_grad=True

perform a mathematical operation

compute gradient using .backward()

Explain why autograd is important for neural networks.

#Write answer of Question 3

In [370]:
a = torch.tensor(5.0, requires_grad=True)
b = a**3 - 2*a + 2
b.backward()

print(f"Tensor a: {a}")
print(f"Result b: {b}")
print(f"Gradient of b with respect to a (db/da): {a.grad}")

Tensor a: 5.0
Result b: 117.0
Gradient of b with respect to a (db/da): 73.0


Importance of Autograd :
- It is the only tool that can calculate gradients automatically which powers the neural network training.
- Neural network use it for Backpropagation where weights and biases are modified according to the aclculated loss with respect to the gradients.
- It is highly efficient for calculating large complex neural network where manually calculating the gradients is almost impossible or error prone.
- It manages a dynamic computational graph which is built during each forward pass, later it helps easy debugging.



# Question 4

Create a custom Dataset class and use DataLoader.

Explain the role of:

Dataset

DataLoader

batch_size

shuffle

#  Write answer of Question 4

In [371]:
class IrishCustomDataset(Dataset):
    def __init__(self, features, labels):
        self.features = torch.tensor(features, dtype=torch.float32)
        self.labels = torch.tensor(labels, dtype=torch.long)

    def __len__(self):
      return len(self.features)

    def __getitem__(self, idx):
      return self.features[idx], self.labels[idx]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

train_data = IrishCustomDataset(X_train, y_train)
test_data = IrishCustomDataset(X_test, y_test)

batch_sz = 16
train_loader = DataLoader(train_data, batch_size=batch_sz, shuffle=True)
test_loader = DataLoader(test_data, batch_size=batch_sz, shuffle=False)

print(f"training samples: {len(train_data)}")
print(f"testing samples: {len(test_data)}")


training samples: 120
testing samples: 30


The role of the dataset is to encapsulate the data and provide an interface for accessing individual data points, it abstracts away the details of how the data is stored or loaded from the disk, presenting a uniform way for dataloader to consume it.

the dataloader wraps an iterable around the dataset to provide an efficient and convenient way to iterate over the data in mini batches, it handles the important aspects like batching, shuffling, parallel loading.

Batch size determines the number of samples grouped together to form a mini batch, as example if the batch size is 16 then tthe dataloader will create the tensors containing 16 samples at a time. Choosing the appropriate batch size is most important bacause large size can take large memory and slow training as well as create generalization issues, when smaller size can lead to noisier gradient estimates, which might help escape local minima and improve generalization but training can be slower and gradients might fluctuate more.

Keeping shuffle true randomly order the data in each epoach, which helps the model by preventing memorizing the order of samples and proovides better generalization, preventing overfitting.

# Question 5

Design an Artificial Neural Network (ANN) using torch.nn.Module.

Your model must include:

input layer

at least one hidden layer

output layer

Explain the purpose of:

nn.Linear

activation functions

loss function

# Write answer of Question 5

In [372]:
class IrishNN(nn.Module):
  def __init__(self, num_features):
    super().__init__()

    self.model = nn.Sequential(
        nn.Linear(num_features, 128),
        nn.ReLU(),
        nn.Linear(128, 64),
        nn.ReLU(),
        nn.Linear(64, num_classes)
    )

  def forward(self, X):
    return self.model(X)



epochs = 20
lr = 0.1

model = IrishNN(X_train.shape[1])
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=lr)

print(model.parameters)

# print(len(train_loader))

<bound method Module.parameters of IrishNN(
  (model): Sequential(
    (0): Linear(in_features=4, out_features=128, bias=True)
    (1): ReLU()
    (2): Linear(in_features=128, out_features=64, bias=True)
    (3): ReLU()
    (4): Linear(in_features=64, out_features=3, bias=True)
  )
)>



*   **`nn.Linear`**: `nn.Linear(num_features, 128)` transforms 4 input features into 128 output values.
*   **Activation Functions (`nn.ReLU`)**: Introduce non-linearity into the network, allowing it to learn complex patterns. `ReLU(x) = max(0, x)` is used for its simplicity and to prevent vanishing gradients.
*   **Loss Function (`nn.CrossEntropyLoss`)**: Quantifies the difference between model predictions and true labels. It's suitable for multi-class classification.

# Question 6

Implement a complete PyTorch training pipeline that includes:

Forward pass

Loss calculation

Backpropagation

Optimizer step

Clearing gradients

Train the model for multiple epochs.

#  Write answer of Question 6

In [373]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)

for epoch in range(epochs):
  total_epoch_loss=0
  for batch_features, batch_labels in train_loader:

    # forward pass
    output=model(batch_features)

    # loss calculation
    loss=criterion(output, batch_labels)
    optimizer.zero_grad()

    # backward pass
    loss.backward()

    # update
    optimizer.step()
    # print(loss.item())
    total_epoch_loss = total_epoch_loss + loss.item()

avg_loss= total_epoch_loss/len(train_loader)

print(avg_loss)

0.3061929028481245


1.  **`optimizer.zero_grad()`**: Clears old gradients from previous steps.
2.  **`outputs = model(inputs)`**: Performs the forward pass, getting predictions.
3.  **`loss = criterion(outputs, labels)`**: Calculates the loss between predictions and actual labels.
4.  **`loss.backward()`**: Computes gradients for all parameters.
5.  **`optimizer.step()`**: Updates model weights using the calculated gradients.

# Question 7

Evaluate your model on the test dataset and compute:

accuracy

loss

Explain whether the model is underfitting or overfitting.

#  Write answer of Question 7

In [374]:
model.eval()
correct = 0
total = 0

with torch.no_grad(): # Disable gradient computation
    for batch_features, batch_labels in test_loader:
        outputs = model(batch_features)
        _,predicted = torch.max(outputs,1)
        # print(predicted)

        # total
        total = total+batch_labels.shape[0]
        correct= correct+(predicted==batch_labels).sum().item()


accuracy = correct / total
print(f"Test Accuracy: {accuracy:.2f}")
print(f"Correct Predictions: {correct}")
print(f"Total Predictions: {total}")

Test Accuracy: 0.97
Correct Predictions: 29
Total Predictions: 30


The IrishNN model achieved a Test Accuracy of 0.97 and an Average Training Loss of 0.306. This result indicates that the model is performing very well.

The model is neither underfitting nor overfitting:

- Not Underfitting: The high accuracy of 97% clearly shows the model has learned the patterns in the data effectively and is not too simple.

- Not Overfitting: The high test accuracy also indicates that the model generalizes well to unseen data and has not simply memorized the training set. If it were overfitting, we would expect a significantly lower test accuracy despite a low training loss.

# Question 8

Modify the architecture and compare performance:

Example modifications:

increase hidden neurons

add another hidden layer

change activation function

Explain the effect on performance.

#  Write answer of Question 8

In [381]:
class ModifiedIrishNN(nn.Module):
    def __init__(self, num_features):
        super().__init__()
        self.model = nn.Sequential(
            # Increased hidden neurons
            nn.Linear(num_features, 256),
            nn.ReLU(),
            # Added another hidden layer
            nn.Linear(256, 128),
            nn.ReLU(),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, num_classes)
        )

    def forward(self, X):
        return self.model(X)


modified_model = ModifiedIrishNN(X_train.shape[1])
criterion_modified = nn.CrossEntropyLoss()
optimizer_modified = optim.Adam(modified_model.parameters(), lr=0.1)
print(modified_model.parameters)

<bound method Module.parameters of ModifiedIrishNN(
  (model): Sequential(
    (0): Linear(in_features=4, out_features=256, bias=True)
    (1): ReLU()
    (2): Linear(in_features=256, out_features=128, bias=True)
    (3): ReLU()
    (4): Linear(in_features=128, out_features=64, bias=True)
    (5): ReLU()
    (6): Linear(in_features=64, out_features=3, bias=True)
  )
)>


Now, let's train this `modified_model` using the same training pipeline and observe its performance.

In [382]:
modified_model.to(device)

epochs_modified = 20

for epoch in range(epochs_modified):
    total_epoch_loss_modified = 0
    for batch_features, batch_labels in train_loader:
        # Forward pass
        outputs_modified = modified_model(batch_features)

        # Loss calculation
        loss_modified = criterion_modified(outputs_modified, batch_labels)
        optimizer_modified.zero_grad()

        # Backward pass
        loss_modified.backward()

        # Update
        optimizer_modified.step()
        total_epoch_loss_modified = total_epoch_loss_modified + loss_modified.item()

    avg_loss_modified = total_epoch_loss_modified / len(train_loader)
print(avg_loss_modified)

0.30909481830894947


Finally, let's evaluate the `modified_model` on the test set and compare its accuracy to the original model.

In [384]:
modified_model.eval()
correct_modified = 0
total_modified = 0

with torch.no_grad(): # Disable gradient computation
    for batch_features, batch_labels in test_loader:
        outputs_modified = modified_model(batch_features)
        _, predicted_modified = torch.max(outputs_modified, 1)

        total_modified += batch_labels.shape[0]
        correct_modified += (predicted_modified == batch_labels).sum().item()

accuracy_modified = correct_modified / total_modified
print(f"Test Accuracy (Modified Model): {accuracy_modified:.2f}")
print(f"Correct Predictions (Modified Model): {correct_modified}")
print(f"Total Predictions (Modified Model): {total_modified}")

# Compare with original model's accuracy
print(f"\nOriginal Model Test Accuracy: {accuracy:.2f}")
print(f"Modified Model Test Accuracy: {accuracy_modified:.2f}")

if accuracy_modified > accuracy:
    print("The Modified Model performed better.")
elif accuracy_modified < accuracy:
    print("The Original Model performed better.")
else:
    print("Both models performed similarly.")

Test Accuracy (Modified Model): 0.97
Correct Predictions (Modified Model): 29
Total Predictions (Modified Model): 30

Original Model Test Accuracy: 0.97
Modified Model Test Accuracy: 0.97
Both models performed similarly.


we modified the IrishNN by increasing hidden neurons (256, 128, 64) and adding an extra hidden layer. We also switched the optimizer from SGD to Adam (with lr=0.1).

Performance Comparison:

Original Model Test Accuracy: 0.97 (Training Loss: 0.306)
Modified Model Test Accuracy: 0.97 (Training Loss: 0.309)
Both models performed similarly. The architectural modifications and optimizer change did not make any improvement because the original model was already performing optimally for the relatively simple Iris dataset.